src/transformers/generation_utils.py

```diff
@@ -89,11 +89,24 @@ def postprocess_next_token_scores(
                 scores[i, banned_tokens] = -float("inf")
 
         if bad_words_ids is not None:
+            # Exclude EOS token (already processed)
+            bad_words_ids = list(filter(lambda bad_token_seq: bad_token_seq != [eos_token_id], bad_words_ids))
             # calculate a list of banned tokens according to bad words
-            banned_tokens = calc_banned_bad_words_ids(input_ids, bad_words_ids)
-
-            for i, banned_tokens in enumerate(banned_tokens):
-                scores[i, banned_tokens] = -float("inf")
+            banned_tokens = calc_banned_bad_words_ids(input_ids.tolist(), bad_words_ids)
+            banned_mask_list = []
+            for idx, batch_banned_tokens in enumerate(banned_tokens):
+                for token in batch_banned_tokens:
+                    banned_mask_list.append([idx, token])
+            if len(banned_mask_list) > 0:
+                banned_mask = torch.LongTensor(banned_mask_list)
+                indices = torch.ones(len(banned_mask))
+                banned_mask = (
+                    torch.sparse.LongTensor(banned_mask.t(), indices, scores.size())
+                    .to_dense()
+                    .bool()
+                    .to(scores.device)
```

Comment:
Not 100% understand what's going on here

Reply:
why not just make a dense tensor to start?

Reply:
Oh I kinda understand now after reading torch.sparse document

Reply:
This is as far as I know the most effective way to create a mask tensor from a list of [idx, pos]. I am not aware of a method to create a dense tensor populated with ones at specific locations, e.g. 
```
[ 0  1  1 ]
[ 0  0  0 ]
[ 1  0  0 ]
```
from: `[[0, 1], [0, 2], [2, 0]]`. As far as I know a dense matrix approach would require a zeroes initialization followed by a for loop to modify the tensor element by element.

Reply:
Maybe a little comment would help people better understand!